src/Components/WebAssembly/JSInterop/src/WebAssemblyJSRuntime.cs

```diff
@@ -61,15 +61,11 @@ protected override void BeginInvokeJS(long asyncHandle, string identifier, strin
         [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "TODO: This should be in the xml suppressions file, but can't be because https://github.com/mono/linker/issues/2006")]
         protected override void EndInvokeDotNet(DotNetInvocationInfo callInfo, in DotNetInvocationResult dispatchResult)
         {
-            // For failures, the common case is to call EndInvokeDotNet with the Exception object.
-            // For these we'll serialize as something that's useful to receive on the JS side.
-            // If the value is not an Exception, we'll just rely on it being directly JSON-serializable.
-            var resultOrError = dispatchResult.Success ? dispatchResult.Result : dispatchResult.Exception!.ToString();
-
-            // We pass 0 as the async handle because we don't want the JS-side code to
-            // send back any notification (we're just providing a result for an existing async call)
-            var args = JsonSerializer.Serialize(new[] { callInfo.CallId, dispatchResult.Success, resultOrError }, JsonSerializerOptions);
-            BeginInvokeJS(0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", args, JSCallResultType.Default, 0);
+            var resultJsonOrErrorMessage = dispatchResult.Success
+                ? dispatchResult.ResultJson!
+                : dispatchResult.Exception!.ToString();
+            InvokeUnmarshalled<string?, bool, string, object>("Blazor._internal.endInvokeDotNetFromJS",
+                callInfo.CallId, dispatchResult.Success, resultJsonOrErrorMessage);
```

Comment:
Instead of doing marshalled-interop-within-marshalled-interop to return the results, it now uses the lower-level unmarshalled API to return the results. This eliminates some JSON serialization.